sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -439,4 +652,553 @@ public void process(ProcessContext c) {
       }
     }
   }
+
+  /**
+   * Specifies how to write elements to individual files in {@link FileIO#write} and {@link
+   * FileIO#writeDynamic}.
+   */
+  public interface Sink<ElementT> extends Serializable {
+    /** Initializes writing to the given channel. */
+    void open(WritableByteChannel channel) throws IOException;
+
+    /** Appends a single element to the file. */
+    void write(ElementT element) throws IOException;
+
+    /**
+     * Flushes the buffered state (if any) before the channel is closed. Does not need to close the
+     * channel.
+     */
+    void flush() throws IOException;
+  }
+
+  /** Implementation of {@link #write} and {@link #writeDynamic}. */
+  @AutoValue
+  @Experimental(Experimental.Kind.SOURCE_SINK)
+  public abstract static class Write<DestinationT, UserT>
+      extends PTransform<PCollection<UserT>, WriteFilesResult<DestinationT>> {
+    /**
+     * Information available for constructing a filename in {@link FilenamePolicy}. A {@link
+     * FilenamePolicy} MUST return unique filenames for different combinations of {@link
+     * #getNumShards}, {@link #getShardIndex}, {@link #getWindow}, {@link #getPane}.
+     */
+    interface FilenameContext {
+      /** Current window being written. */
+      BoundedWindow getWindow();
+
+      /** Pane of the current window being written. */
+      PaneInfo getPane();
+
+      /** Total number of shards generated for the current window and pane. */
+      int getNumShards();
+
+      /**
+       * Index of the current shard among all the shards generated for the current window and pane.
+       */
+      int getShardIndex();
+
+      /** The specified {@link Compression}. */
+      Compression getCompression();
+    }
+
+    /**
+     * A policy for generating names for shard files using information in {@link FilenameContext}.
+     */
+    interface FilenamePolicy extends Serializable {
```

Comment:
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Won't it get confusing having two FilenamePolicy interfaces?